### PR TITLE
added Function::hessian()

### PR DESCRIPTION
**Pull request by davydden**
btw, i think we should change all other pure functions to `AssertThrow` as they shall not be called anyway.

### CODE REVIEW

File: include/deal.II/base/function.h [hunk 1]
```diff
@@ -310,6 +310,33 @@ class Function : public FunctionTime<Number>,
                                       std::vector<Vector<Number> >   &values) const;
 
   /**
+   * Compute the Hessian of a given component at point <tt>p</tt>.
```
Comment (bangerth) [inline include/deal.II/base/function.h]: Can you add a definition of what the "Hessian" actually is? I don't think it is universally understood as simply the second derivative (unless you are an optimization person, of course).

**Comment by bangerth** [comment]
Looks good. Can you elaborate which functions you want to throw an exception rather than being pure?

**Comment by davydden** [comment]
i wonder if i should force it to be a SymmericTensor? As long as the function is continuous the order of differentiation does not matter.

**Comment by bangerth** [comment]
Yes, nice idea. Make it so!

**Comment by davydden** [comment]
with regards to assertions, i changed function.templates.h to do `AssertThrow` instead of `Assert` in a couple of places.

**Comment by davydden** [comment]
let me check that it actually compiles...

**Comment by bangerth** [comment]
I don't think that's necessary. We use Assert for things that can be checked in debug mode (such as that you call the correct function with the correct arguments) and AssertThrow for things that may depend on run-time input. You're not going to call a different function based on input parameters, so I think Assert is the correct choice.

**Comment by davydden** [comment]
ok, i'll drop that commit.

**Comment by bangerth** [comment]
Thanks. While you're at it, also add an entry to `doc/news/changes.h`.

**Pull request merged**